data/decode.go

```diff
@@ -31,71 +125,117 @@ type NodeEdgeChildren struct {
 //
 // output can also be a *reflect.Value
 func Decode(input NodeEdgeChildren, output interface{}) error {
-	var vOut reflect.Value
-	var tOut reflect.Type
+	vOut := reflect.Indirect(reflect.ValueOf(output))
+	tOut := vOut.Type()
 
-	if reflect.TypeOf(output).String() == "*reflect.Value" {
-		outputV, ok := output.(*reflect.Value)
+	if tOut == reflectValueT {
+		// `output` was a reflect.Value or *reflect.Value
+		vOut = vOut.Interface().(reflect.Value)
+		tOut = vOut.Type()
+	}
+
+	if k := tOut.Kind(); k != reflect.Struct {
+		return fmt.Errorf("Error decoding to %v; must be a struct", k)
+	}
+
+	// Group points and children by type
+	pointGroups := make(map[string]GroupedPoints)
+	edgePointGroups := make(map[string]GroupedPoints)
+	childGroups := make(map[string][]NodeEdgeChildren)
+
+	for _, p := range input.NodeEdge.Points {
+		g, ok := pointGroups[p.Type]
 		if !ok {
-			return errors.New("Error converting interface")
+			g = GroupedPoints{}
 		}
-
-		vOut = outputV.Elem()
-		tOut = outputV.Type().Elem()
-	} else {
-		vOut = reflect.ValueOf(output).Elem()
-		tOut = reflect.TypeOf(output).Elem()
+		if p.Key != "" {
+			g.Keyed = true
+		}
+		if index := int(p.Index); index > g.IndexMax {
+			g.IndexMax = index
+		}
+		g.Points = append(g.Points, p)
+		pointGroups[p.Type] = g
 	}
+	for _, p := range input.NodeEdge.EdgePoints {
+		g, ok := edgePointGroups[p.Type]
+		if !ok {
+			g = GroupedPoints{}
+		}
+		if p.Key != "" {
+			g.Keyed = true
+		}
 
-	pointValues := make(map[string]reflect.Value)
-	edgeValues := make(map[string]reflect.Value)
-	childValues := make(map[string]reflect.Value)
+		if index := int(p.Index); index > g.IndexMax {
+			g.IndexMax = index
+		}
+		g.Points = append(g.Points, p)
+		edgePointGroups[p.Type] = g
+	}
+	for _, c := range input.Children {
+		childGroups[c.NodeEdge.Type] = append(childGroups[c.NodeEdge.Type], c)
+	}
 
 	for i := 0; i < tOut.NumField(); i++ {
 		sf := tOut.Field(i)
 		if pt := sf.Tag.Get("point"); pt != "" {
-			pointValues[pt] = vOut.Field(i)
+			g := pointGroups[pt]
+			// Write points into struct field
+			err := setValueFromPointGroup(g, vOut.Field(i))
+			if err != nil {
+				log.Printf("decode error for type %v: %v", pt, err)
+			}
 		} else if et := sf.Tag.Get("edgepoint"); et != "" {
-			edgeValues[et] = vOut.Field(i)
+			g := edgePointGroups[et]
+			// Write points into struct field
+			err := setValueFromPointGroup(g, vOut.Field(i))
+			if err != nil {
+				log.Printf("decode error for type %v: %v", et, err)
```

Comment:
Should these return an error instead of printing to console?

Reply:
if the function returns an error, then yes. 

I also learned recently the fmt.Errorf should use %w for err, as then it wraps the error. You probably know this.